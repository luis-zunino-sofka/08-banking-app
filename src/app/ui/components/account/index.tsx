import { useCreateBankAccount } from "@core/hooks";
import { decryptAES } from "@core/utils";
import React from "react";
import { Toaster } from "react-hot-toast";
import { Button } from "../button";
import "./styles.scss";

export const CreateBankAccount: React.FC = () => {
  const { isLoading, accounts, errors, handleSubmit, register, onSubmit } =
    useCreateBankAccount();

  return (
    <div>
      <form onSubmit={handleSubmit(onSubmit)} className="form">
        <div style={{ display: "flex", gap: "10px" }}>
          <div>
            <div>
              <label htmlFor="accountType">Tipo de Cuenta</label>
              <select
                id="accountType"
                {...register("accountType", {
                  required: "Selecciona un tipo de cuenta",
                })}
              >
                <option key={1} value="">
                  Selecciona el tipo de cuenta
                </option>
                <option key={2} value="corriente">
                  Corriente
                </option>
              </select>
              <span className="error-message">
                {errors.accountType && errors.accountType.message}
              </span>
            </div>
            <div>
              <label htmlFor="initialBalance">Saldo Inicial</label>
              <input
                type="number"
                id="initialBalance"
                placeholder="Saldo inicial en $"
                min={0}
                {...register("initialBalance", {
                  required: "El saldo inicial es obligatorio",
                  min: {
                    value: 0,
                    message: "El saldo inicial no puede ser negativo",
                  },
                })}
              />
              <span className="error-message">
                {errors.initialBalance && errors.initialBalance.message}
              </span>
            </div>
            <Button label="Crear Cuenta" isLoading={isLoading} />
          </div>
          <div className="user-accounts">
            <p>Cuentas del usuario:</p>
            {accounts.map((a) => (
              <div key={a.accountId} className="user-accounts__item">
                <div className="user-accounts__info">
                  <span>NÂº de cuenta:</span> {decryptAES(a.encryptedNumber)}
                </div>
                <div className="user-accounts__info amount">
                  <span>Monto:</span> $ {a.amount}
                </div>
              </div>
            ))}
          </div>
        </div>
      </form>
      <Toaster />
    </div>
  );
};
