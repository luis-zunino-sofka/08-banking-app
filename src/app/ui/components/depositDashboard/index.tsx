import { useDeposit } from "@core/hooks";
import { decryptAES } from "@core/utils";
import { Toaster } from "react-hot-toast";
import { Button } from "../button";

export const DepositDashboard = () => {
  const {
    isLoading,
    accounts,
    errors,
    isAccountSelected,
    handleSubmit,
    register,
    onSubmit,
  } = useDeposit();

  return (
    <div className="deposit">
      <form
        id="deposit-form"
        className="form"
        onSubmit={handleSubmit(onSubmit)}
      >
        <label htmlFor="deposit-source">
          Selecciona el origen del depósito:
        </label>
        <select
          id="deposit-source"
          {...register("depositSource", {
            required: "El monto a depositar es obligatorio",
          })}
        >
          <option value="branch-deposit" defaultValue="branch-deposit">
            Depósito desde sucursal (Sin costo)
          </option>
          <option value="atm">
            Depósito desde cajero automático ($2 de costo)
          </option>
          <option value="account">
            Depósito desde otra cuenta ($1.5 de costo)
          </option>
        </select>
        {errors.depositSource && (
          <span className="error-message">{errors.depositSource.message}</span>
        )}
        <label htmlFor="deposit-amount">Monto:</label>
        <input
          type="number"
          id="deposit-amount"
          placeholder="Monto a depositar"
          required
          min={0}
          {...register("amount", {
            required: "El monto a depositar es obligatorio",
            min: {
              value: 1,
              message: "El monto a depositar debe ser mayor a 0",
            },
          })}
        />
        {errors.amount && (
          <span className="error-message">{errors.amount.message}</span>
        )}
        <div>
          <label htmlFor="accountId">
            {isAccountSelected
              ? "Cuenta de extracción:"
              : "Cuenta a depositar:"}
          </label>
          <select
            id="accountId"
            {...register("accountId", {
              required: "La cuenta bancaria es obligatorio",
            })}
            required
          >
            <option value="">Elige una cuenta</option>
            {accounts.map((a) => (
              <option key={a.accountId} value={a.accountId}>
                Cuenta: {decryptAES(a.encryptedNumber)} - Monto: {a.amount}
              </option>
            ))}
          </select>

          {errors.accountNumberToDeposit && (
            <span className="error-message">
              {errors.accountNumberToDeposit.message}
            </span>
          )}
        </div>
        {isAccountSelected && (
          <>
            <label htmlFor="account-number-to-deposit">
              Selecciona la cuenta a depositar:
            </label>
            <input
              type="text"
              id="account-number-to-deposit"
              placeholder="Cuenta a depositar"
              required
              min={0}
              {...register("accountNumberToDeposit", {
                required: "La cuenta a depositar es obligatorio",
              })}
            />

            {errors.accountNumberToDeposit && (
              <span className="error-message">
                {errors.accountNumberToDeposit.message}
              </span>
            )}
          </>
        )}
        <Button label="Depositar" isLoading={isLoading} />
        <p>* El costo de la transacción se deducirá del saldo.</p>
      </form>
      <Toaster />
    </div>
  );
};
