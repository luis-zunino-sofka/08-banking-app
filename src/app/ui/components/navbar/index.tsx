import "./styles.scss";
import { INavItem, IState } from "@core/interfaces";
import { FC } from "react";

export interface IDashboardProps {
  state: IState;
  isRefechingAccounts: boolean;
  navItems: INavItem[];
  refetchAccounts: () => void;
  handleLogout: () => void;
}

export const Navbar: FC<IDashboardProps> = (props) => {
  const {
    state,
    navItems,
    isRefechingAccounts,
    refetchAccounts,
    handleLogout,
  } = props;

  return (
    <nav className="navbar">
      <div className="navbar__title">
        <h1>Bienvenido {state.loginData?.username}</h1>
      </div>
      <div className="navbar__balance">
        <span>Saldo disponible: </span>
        <strong>${state.balance}</strong>
        <button
          disabled={isRefechingAccounts}
          className="navbar__reload"
          onClick={() => refetchAccounts()}
        >
          Actualizar
        </button>
      </div>
      <div className="navbar__nav">
        <ul className="navbar__list">
          {navItems.map((item) => (
            <li key={item.path} className="navbar-item">
              <button
                className={`navbar__link ${
                  item.isActive ? "navbar__link--active" : ""
                }`}
                onClick={item.onClick}
              >
                {item.label}
              </button>
            </li>
          ))}
          <li className="navbar__item">
            <button
              className="navbar__link navbar__logout"
              onClick={handleLogout}
            >
              Cerrar sesi√≥n
            </button>
          </li>
        </ul>
      </div>
    </nav>
  );
};
