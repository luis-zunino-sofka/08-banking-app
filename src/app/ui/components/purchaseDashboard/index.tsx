import { Toaster } from "react-hot-toast";
import { usePurchase } from "@core/hooks";
import { decryptAES } from "@core/utils";
import { Button } from "../button";
import "../../styles/styles.scss";

export const PurchaseDashboard = () => {
  const { isLoading, accounts, errors, register, handleSubmit, onSubmit } =
    usePurchase();

  return (
    <div>
      <form
        id="purchase-form"
        className="form"
        onSubmit={handleSubmit(onSubmit)}
      >
        <div>
          <label htmlFor="purchase-type">Selecciona el tipo de compra:</label>
          <select
            id="purchase-type"
            {...register("purchaseType", {
              required: "Por favor selecciona un tipo de compra",
            })}
          >
            <option value="">Elige el tipo de compra</option>
            <option value="physical">Establecimiento físico (Sin costo)</option>
            <option value="online">En línea (Costo de seguro de $5 USD)</option>
          </select>
          {errors.purchaseType && (
            <span className="error-message">{errors.purchaseType.message}</span>
          )}
        </div>
        <div>
          <label htmlFor="account-number-to-deposit">
            Selecciona la cuenta a depositar:
          </label>
          <select
            id="account-number-to-deposit"
            {...register("accountId", {
              required: "La cuenta bancaria a depositar es obligatorio",
            })}
          >
            <option value="">Elige una cuenta</option>
            {accounts.map((a) => (
              <option key={a.accountId} value={a.accountId}>
                Cuenta: {decryptAES(a.encryptedNumber)} - Monto: {a.amount}
              </option>
            ))}
          </select>

          {errors.accountId && (
            <span className="error-message">{errors.accountId.message}</span>
          )}
        </div>
        <div>
          <label htmlFor="purchase-amount">
            Ingresa el monto a depositar:
            <input
              type="number"
              id="purchase-amount"
              placeholder="Monto de la compra en $"
              {...register("purchaseAmount", {
                required: "El monto es obligatorio",
                min: {
                  value: 1,
                  message: "El monto debe ser mayor que $0",
                },
              })}
            />
          </label>
          {errors.purchaseAmount && (
            <span className="error-message">
              {errors.purchaseAmount.message}
            </span>
          )}
        </div>
        <Button label="Comprar" isLoading={isLoading} />
      </form>

      <Toaster />
    </div>
  );
};
