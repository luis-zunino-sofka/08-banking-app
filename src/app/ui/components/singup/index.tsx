import { useSignup } from "@core/hooks/signup/useSignup";
import { ROUTE_PATH } from "@core/constants";
import { Toaster } from "react-hot-toast";
import "./styles.scss";
import { InputSignup } from "./InputSignup";

export const SignupForm = () => {
  const { errors, isLoading, onSubmit, register, handleSubmit, watch } =
    useSignup();

  return (
    <div className="signup-form">
      <form
        id="register-form"
        className="signup-form__form"
        onSubmit={handleSubmit(onSubmit)}
      >
        <h2 className="signup-form__title">Regístrate</h2>
        <div className="flex">
          <InputSignup
            register={register("name", {
              required: "El nombre es obligatorio",
            })}
            name="name"
            type="text"
            label="Nombre:"
            errors={errors}
            placeHolder="Ingrese su nombre"
          />

          <InputSignup
            register={register("lastName", {
              required: "El apellido es obligatorio",
            })}
            name="lastName"
            type="text"
            label="Apellido:"
            errors={errors}
            placeHolder="Ingrese su apellido"
          />
        </div>
        <div className="flex">
          <InputSignup
            register={register("dni", { required: "El DNI es obligatorio" })}
            name="dni"
            type="text"
            label="DNI:"
            errors={errors}
            placeHolder="Ingrese su dni"
          />

          <InputSignup
            register={register("username", {
              required: "El nombre de usuario es obligatorio",
            })}
            name="username"
            type="text"
            label="Nombre de usuario:"
            errors={errors}
            placeHolder="Ingrese su nombre de usuario"
          />
        </div>
        <div className="flex">
          <InputSignup
            register={register("password", {
              required: "El nombre de usuario es obligatorio",
            })}
            name="password"
            type="password"
            label="Contraseña:"
            errors={errors}
            placeHolder="Contraseña"
          />

          <InputSignup
            register={register("confirmPassword", {
              required: "La confirmación de la contraseña es obligatoria",
              validate: (value) =>
                value === watch("password") || "Las contraseñas no coinciden",
            })}
            name="confirmPassword"
            type="password"
            label="Confirmar contraseña:"
            errors={errors}
            placeHolder="Confirmar contraseña"
          />
        </div>
        <button
          type="submit"
          className="signup-form__submit"
          disabled={isLoading}
        >
          Registrarse
        </button>
        <div className="signup-form__footer">
          <p>
            ¿Ya tienes cuenta?
            <a
              href={ROUTE_PATH.LOGIN}
              id="login-link"
              className="signup-form__login-link"
            >
              Inicia sesión aquí
            </a>
          </p>
        </div>
      </form>
      <Toaster />
    </div>
  );
};
