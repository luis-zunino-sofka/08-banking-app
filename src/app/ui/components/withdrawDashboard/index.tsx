import { Toaster } from "react-hot-toast";
import { decryptAES } from "@core/utils";
import { useWithdrawDashboard } from "@core/hooks";

export const WithdrawDashboard = () => {
  const { accounts, register, handleSubmit, errors, onSubmit } =
    useWithdrawDashboard();

  return (
    <div className="withdraw">
      <form
        id="withdraw-form"
        className="form"
        onSubmit={handleSubmit(onSubmit)}
      >
        <div>
          <label htmlFor="withdraw-amount">Monto:</label>
          <input
            type="number"
            id="withdraw-amount"
            placeholder="Monto a retirar en $"
            {...register("amount", {
              required: "El monto es obligatorio.",
              min: {
                value: 0,
                message: "El monto debe ser mayor que 0.",
              },
            })}
          />
          {errors.amount && (
            <span className="error-message">{errors.amount.message}</span>
          )}
          <>
            <label htmlFor="account-number-to-withdraw">
              Selecciona la cuenta a retirar:
            </label>
            <select
              id="account-number-to-withdraw"
              {...register("accountNumberToWithdraw", {
                required: "La cuenta bancaria a retirar es obligatorio",
              })}
              required
            >
              <option value="">Elige una cuenta</option>
              {accounts.map((a) => (
                <option key={a.accountId} value={a.accountId}>
                  Cuenta: {decryptAES(a.encryptedNumber)} - Monto: {a.amount}
                </option>
              ))}
            </select>

            {errors.accountNumberToWithdraw && (
              <span className="error-message">
                {errors.accountNumberToWithdraw.message}
              </span>
            )}
          </>
        </div>

        <p>* El retiro en cajero tiene un costo de $1 USD por transacci√≥n.</p>
        <button type="submit">Retirar</button>
      </form>
      <Toaster />
    </div>
  );
};
