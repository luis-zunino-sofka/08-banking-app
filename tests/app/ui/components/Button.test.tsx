import { render, screen } from "@testing-library/react";
import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import React from "react";
import { Button } from "../../../../src/app/ui/components/button";
import "@testing-library/jest-dom";

describe("Button Component", () => {
  beforeEach(() => {
    vi.mock("react-spinner", () => ({}));
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  test("Match snapshoot", () => {
    const { asFragment } = render(
      <Button label="Iniciar sesión" isLoading={false} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("debe renderizar el texto del botón cuando isLoading es false", () => {
    render(<Button label="Iniciar sesión" isLoading={false} />);
    const button = screen.getByRole("button");
    expect(button).toHaveTextContent("Iniciar sesión");

    const loader = screen.queryByTestId("loader");
    expect(loader).not.toBeInTheDocument();
  });

  test("debe renderizar el spinner cuando isLoading es true", () => {
    render(<Button label="Iniciar sesión" isLoading={true} />);
    const loader = screen.getByTestId("loader");
    expect(loader).toBeInTheDocument();

    const button = screen.getByRole("button");
    expect(button).not.toHaveTextContent("Iniciar sesión");
  });

  test("debe mostrar el texto del botón y no el spinner cuando isLoading es false", () => {
    render(<Button label="Enviar" isLoading={false} />);

    const button = screen.getByRole("button");
    expect(button).toHaveTextContent("Enviar");
    expect(button).not.toHaveTextContent("Cargando");
    expect(screen.queryByTestId("loader")).not.toBeInTheDocument();
  });

  test("debe mostrar el spinner y no el texto del botón cuando isLoading es true", () => {
    render(<Button label="Enviar" isLoading={true} />);

    const button = screen.getByRole("button");
    expect(button).not.toHaveTextContent("Enviar");
    expect(screen.getByTestId("loader")).toBeInTheDocument();
  });
});
