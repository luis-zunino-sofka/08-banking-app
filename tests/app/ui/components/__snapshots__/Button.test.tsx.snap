// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > Match snapshoot 1`] = `
<DocumentFragment>
  <button
    class="button"
    type="submit"
  >
    Iniciar sesi√≥n
  </button>
</DocumentFragment>
`;
